File: CashuWallet/Core/TokenParser.swift
import Foundation
import CashuDevKit

enum TokenParser {
    static func normalizedToken(from rawToken: String) -> String? {
        let token = rawToken.trimmingCharacters(in: .whitespacesAndNewlines)
        guard isCashuToken(token) else { return nil }
        return token
    }

    static func isCashuToken(_ token: String) -> Bool {
        token.lowercased().hasPrefix("cashu")
    }

    static func isCashuDeepLinkToken(_ token: String) -> Bool {
        let lowercased = token.lowercased()
        return lowercased.hasPrefix("cashua") || lowercased.hasPrefix("cashub")
    }

    static func tokenInfo(from tokenString: String) -> TokenInfo? {
        guard isCashuToken(tokenString),
              let token = try? Token.decode(encodedToken: tokenString),
              let mint = try? token.mintUrl().url,
              let proofs = try? token.proofsSimple() else {
            return nil
        }

        let amount = proofs.reduce(UInt64(0)) { $0 + $1.amount.value }
        return TokenInfo(
            amount: amount,
            mint: mint,
            unit: "sat",
            memo: token.memo(),
            proofCount: proofs.count
        )
    }
}

File: android/app/src/main/java/org/cashu/wallet/Core/TokenParser.kt
package org.cashu.wallet.Core

import org.cashu.wallet.Models.TokenInfo
import org.cashudevkit.CurrencyUnit as CdkCurrencyUnit
import org.cashudevkit.Token as CdkToken

object TokenParser {
    private val tokenPrefixes = listOf("cashuA", "cashuB", "cashuC")

    fun extractToken(raw: String): String? {
        val trimmed = raw.trim()
        val withoutScheme = when {
            trimmed.startsWith("cashu://", ignoreCase = true) -> trimmed.drop("cashu://".length)
            trimmed.startsWith("cashu:", ignoreCase = true) -> trimmed.drop("cashu:".length)
            else -> trimmed
        }
        return withoutScheme.takeIf { tokenPrefixes.any { prefix -> it.startsWith(prefix, ignoreCase = true) } }
    }

    fun isCashuToken(raw: String): Boolean = extractToken(raw) != null

    fun tokenInfo(from: String): TokenInfo? {
        val token = extractToken(from) ?: return null
        val decoded = runCatching { CdkToken.decode(token) }.getOrNull() ?: return null
        val proofs = runCatching { decoded.proofsSimple() }.getOrDefault(emptyList())
        return TokenInfo(
            amount = runCatching { decoded.value().value.toLong() }.getOrDefault(0),
            mint = runCatching { decoded.mintUrl().url }.getOrDefault("Unknown mint"),
            unit = decoded.unit()?.toDomainUnit() ?: "sat",
            memo = decoded.memo(),
            proofCount = proofs.size,
        )
    }

    fun p2pkPubkeys(from: String): List<String> {
        val token = extractToken(from) ?: return emptyList()
        val decoded = runCatching { CdkToken.decode(token) }.getOrNull() ?: return emptyList()
        return runCatching { decoded.p2pkPubkeys() }.getOrDefault(emptyList())
    }

    private fun CdkCurrencyUnit.toDomainUnit(): String = when (this) {
        CdkCurrencyUnit.Sat -> "sat"
        CdkCurrencyUnit.Msat -> "msat"
        CdkCurrencyUnit.Usd -> "usd"
        CdkCurrencyUnit.Eur -> "eur"
        CdkCurrencyUnit.Auth -> "auth"
        is CdkCurrencyUnit.Custom -> unit
    }
}
